### PR TITLE
Enabling the retrieval of whisper's hidden states

### PR DESCRIPTION
**Pull request by Hguimaraes**
Hello guys,

I'm trying to extend the Whisper class to enable the retrieval of the Hidden States. I think it is handy for people working with speech representation learning. It is a small change. Also, I tried to make it compatible with the previous version, so it should not break any examples or codes that are already using this class.

Best regards,
Heitor

### CODE REVIEW

File: speechbrain/lobes/models/huggingface_whisper.py [hunk 1]
```diff
@@ -65,13 +65,15 @@ def __init__(
         freeze=False,
         freeze_encoder=False,
         output_attentions=True,
+        output_hidden_states=False,
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/huggingface_whisper.py:68]: I'd prefer to see "output_all_hiddens" instead of "output_hidden_states" as we already have something really similar to what you did with Wav2vec2 and we named it "output_all_hiddens" see: https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/lobes/models/huggingface_wav2vec.py

**Review by Adel-Moumen — CHANGES_REQUESTED** [review]
Hello, 

Many thanks for the PR! 

I left one comment on the naming of the variable + docstring. After that, I think everything is LGTM.

File: speechbrain/lobes/models/huggingface_whisper.py (hunk 1 above)
Comment (Hguimaraes) [reply speechbrain/lobes/models/huggingface_whisper.py]: Thank you for your answer! I pushed the modifications. I changed the variable name, inserted it on the docstring, and also stack the tensors instead of returning a tuple... Now it is more similar to the wav2vec2 class.

File: speechbrain/lobes/models/huggingface_whisper.py [hunk 2]
```diff
@@ -131,18 +138,29 @@ def forward(self, wav, decoder_input_ids=None):
                 out_encoder = self.forward_encoder(wav)
                 if self.encoder_only:
                     return out_encoder
-                logits, attn = self.forward_decoder(
-                    out_encoder, decoder_input_ids
-                )
+
+                if self.output_all_hiddens:
+                    logits, attn = self.forward_decoder(
+                        out_encoder[-1, ...], decoder_input_ids
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/huggingface_whisper.py:144]: why is there now `out_encoder[-1, ...]` instead of `out_encoder[-1]` ?

File: speechbrain/lobes/models/huggingface_whisper.py [hunk 3]
```diff
@@ -155,10 +173,16 @@ def forward_encoder(self, wav):
...
-                return self.model.encoder(mel).last_hidden_state
+                return self._get_encoder_states(wav)
+        else:
+            return self._get_encoder_states(wav)
+
+    def _get_encoder_states(self, wav):
+        mel = self._get_mel(wav)
+        if self.output_all_hiddens:
+            states = self.model.encoder(mel, output_hidden_states=True)
+            return torch.stack(states.hidden_states)
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/huggingface_whisper.py:184]: is it the same to `torch.stack(list(states.hidden_states), dim=0)` (taken from wav2vec2)?

Comment (Hguimaraes) [reply speechbrain/lobes/models/huggingface_whisper.py:184]: Yes! I think it is not necessary to take this extra step to first convert to a list, we can stack the tuples directly.

File: speechbrain/lobes/models/huggingface_whisper.py (hunk 2 above)
Comment (Hguimaraes) [reply speechbrain/lobes/models/huggingface_whisper.py:144]: Because now I'm not returning a tuple of tensors as before, I'm returning the stacked tensor (num_layers, B, T, C) [As I answered you in the other comment]

**Comment by Hguimaraes** [comment]
Hi @Adel-Moumen,

Please let me know if there is anything else I could do on this PR. I think I have answered the previous questions and the interface is similar to the wav2vec2 class.

Best regards,

**Comment by Adel-Moumen** [comment]
> Hi @Adel-Moumen,
> 
> Please let me know if there is anything else I could do on this PR. I think I have answered the previous questions and the interface is similar to the wav2vec2 class.
> 
> Best regards,

Hello @Hguimaraes, 

I will finish the review this week :-)

File: speechbrain/lobes/models/huggingface_whisper.py [hunk 4]
```diff
@@ -155,10 +173,16 @@ def forward_encoder(self, wav):
 
         if self.freeze_encoder:
             with torch.no_grad():
-                mel = self._get_mel(wav)
-                return self.model.encoder(mel).last_hidden_state
+                return self._get_encoder_states(wav)
+        else:
+            return self._get_encoder_states(wav)
+
+    def _get_encoder_states(self, wav):
```
Comment (Adel-Moumen) [inline speechbrain/lobes/models/huggingface_whisper.py:180]: please add some doc string on this function

File: speechbrain/lobes/models/huggingface_whisper.py (hunk 2 above)
Comment (Adel-Moumen) [inline speechbrain/lobes/models/huggingface_whisper.py:144]: From my understanding, `out_encoder[-1, ...]` is exactly the same as `out_encoder[-1]`, so why adding "..." ?

**Review by Adel-Moumen — CHANGES_REQUESTED** [review]
Hello, 

Thanks again for your great work! Please take a look at my latest comments :-) 

Btw, I wish you a great end of the year :-)

**Comment by Hguimaraes** [comment]
Hi @Adel-Moumen,

Implemented the changes. If there is anything else, just let me know, and I can update it.
I hope you have a great end of the year as well!

Best,

**Review by Adel-Moumen — APPROVED** [review]
Everything sounds good! Many thanks for the PR! :-)
